core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4

```diff
@@ -841,6 +856,121 @@ pathSpecification
     : pathElement (',' pathElement)*
     ;
 
+functionSpecification
+    : FUNCTION qualifiedName parameterDeclarationList
+        returnsClause routineCharacteristic*
+        (returnStatement | compoundStatement)
+    ;
+
+parameterDeclarationList
+    : '(' (parameterDeclaration (',' parameterDeclaration)*)? ')'
+    ;
+
+parameterDeclaration
+    : identifier? type
+    ;
+
+returnsClause
+    : RETURNS type
+    ;
+
+routineCharacteristic
+    : LANGUAGE identifier               #languageCharacteristic
+    | NOT? DETERMINISTIC                #deterministicCharacteristic
+    | RETURNS NULL ON NULL INPUT        #returnsNullOnNullInputCharacteristic
+    | CALLED ON NULL INPUT              #calledOnNullInputCharacteristic
+    | SECURITY (DEFINER | INVOKER)      #securityCharacteristic
+    | COMMENT string                    #commentCharacteristic
+    ;
+
+routineStatement
+    : returnStatement
+    | assignmentStatement
+    | simpleCaseStatement
+    | searchedCaseStatement
+    | ifStatement
+    | iterateStatement
+    | leaveStatement
+    | compoundStatement
+    | loopStatement
+    | whileStatement
+    | repeatStatement
+    ;
+
+returnStatement
+    : RETURN valueExpression
+    ;
+
+assignmentStatement
+    : SET identifier EQ expression
+    ;
+
+simpleCaseStatement
+    : CASE expression caseStatementWhenClause+ elseClause? END CASE
+    ;
+
+searchedCaseStatement
+    : CASE caseStatementWhenClause+ elseClause? END CASE
+    ;
+
+caseStatementWhenClause
+    : WHEN expression THEN sqlStatementList
+    ;
+
+ifStatement
+    : IF expression THEN sqlStatementList elseIfClause* elseClause? END IF
+    ;
+
+elseIfClause
+    : ELSEIF expression THEN sqlStatementList
+    ;
+
+elseClause
+    : ELSE sqlStatementList
+    ;
+
+iterateStatement
+    : ITERATE identifier
+    ;
+
+leaveStatement
+    : LEAVE identifier
+    ;
+
+variableDeclaration
+    : DECLARE identifier (',' identifier)* type (DEFAULT valueExpression)?
+    ;
+
+compoundStatement
+    : BEGIN
+      (variableDeclaration SEMICOLON)*
+      sqlStatementList?
+      END
+    ;
+
+loopStatement
+    : (label=identifier ':')? LOOP
+      sqlStatementList
+      END LOOP
+    ;
+
+whileStatement
+    : (label=identifier ':')? WHILE expression DO
+      sqlStatementList
+      END WHILE
```

Comment:
Same here. The spec allows for an optional ending label.

Reply:
Intentionally not supported.